### PR TITLE
Ignore erased scratch bits when constructing `ResultHandle`

### LINKED ISSUE #213: `AssertionError` from `backend.cost()` with classical circuit
Using pytket-quantinuum 0.20.0:

```python
from pytket.circuit import Circuit, Qubit, Bit, OpType, reg_eq
from pytket.extensions.quantinuum import QuantinuumBackend

circuit = Circuit(name="Repetition Code")
data = circuit.add_q_register("data", 3)
ancilla = circuit.add_q_register("anc", 1)
syndrome = circuit.add_c_register("syndrome", 2)
output = circuit.add_c_register("output", 3)

circuit.add_gate(OpType.Reset, ancilla)
circuit.CX(data[0], ancilla[0])
circuit.CX(data[1], ancilla[0])
circuit.Measure(ancilla[0], syndrome[0])
circuit.add_gate(OpType.Reset, ancilla)
circuit.CX(data[1], ancilla[0])
circuit.CX(data[2], ancilla[0])
circuit.Measure(ancilla[0], syndrome[1])
circuit.X(data[0], condition=reg_eq(syndrome, 1))
circuit.X(data[2], condition=reg_eq(syndrome, 2))
circuit.X(data[1], condition=reg_eq(syndrome, 3))
circuit.Measure(data[0], output[0])
circuit.Measure(data[1], output[1])
circuit.Measure(data[2], output[2])

backend = QuantinuumBackend(device_name="H1-1E")
backend.login()

compiled_circuit = backend.get_compiled_circuit(circuit, optimisation_level=1)

print(compiled_circuit.qubits)
print(compiled_circuit.bits)
print(compiled_circuit.get_commands())

cost = backend.cost(compiled_circuit, n_shots=100, syntax_checker="H1-1SC")
print(cost)
```

Produces:
```
Enter your Quantinuum password: 
[q[0], q[1], q[2], q[3]]
[output[0], output[1], output[2], syndrome[0], syndrome[1], tk_SCRATCH_BIT[0], tk_SCRATCH_BIT[1], tk_SCRATCH_BIT[2]]
[Reset q[0];, PhasedX(2.5, 0.5) q[0];, ZZPhase(0.5) q[1], q[0];, ZZPhase(0.5) q[2], q[0];, Rz(0.5) q[1];, PhasedX(0.5, 0.5) q[0];, Measure q[0] --> syndrome[0];, Reset q[0];, PhasedX(2.5, 0.5) q[0];, ZZPhase(0.5) q[2], q[0];, ZZPhase(0.5) q[3], q[0];, Rz(3) q[2];, PhasedX(0.5, 0.5) q[0];, Rz(0.5) q[3];, Measure q[0] --> syndrome[1];, RangePredicate([1,1]) syndrome[0], syndrome[1], tk_SCRATCH_BIT[0];, RangePredicate([2,2]) syndrome[0], syndrome[1], tk_SCRATCH_BIT[1];, RangePredicate([3,3]) syndrome[0], syndrome[1], tk_SCRATCH_BIT[2];, IF ([tk_SCRATCH_BIT[0]] == 1) THEN PhasedX(1, 0) q[1];, IF ([tk_SCRATCH_BIT[2]] == 1) THEN PhasedX(1, 0) q[2];, IF ([tk_SCRATCH_BIT[1]] == 1) THEN PhasedX(1, 0) q[3];, IF ([tk_SCRATCH_BIT[0]] == 1) THEN Phase(0.5);, IF ([tk_SCRATCH_BIT[1]] == 1) THEN Phase(0.5);, IF ([tk_SCRATCH_BIT[2]] == 1) THEN Phase(0.5);, Measure q[1] --> output[0];, Measure q[2] --> output[1];, Measure q[3] --> output[2];]
n_bits = 8
c_bits = [syndrome[1], syndrome[0], output[2], output[1], output[0]]
Traceback (most recent call last):
  File "/home/alec/tmp/pytket-quantinuum/env/lib/python3.10/site-packages/pytket/extensions/quantinuum/backends/quantinuum.py", line 936, in get_result
    return super().get_result(handle)
  File "/home/alec/tmp/pytket-quantinuum/env/lib/python3.10/site-packages/pytket/backends/backend.py", line 347, in get_result
    raise CircuitNotRunError(handle)
pytket.backends.backend_exceptions.CircuitNotRunError: Circuit corresponding to ResultHandle('64c69111e8f44be6a64f71519d93764b', 'null', 8) has not been run by this backend instance.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/alec/r/pytket-quantinuum/costerr.py", line 34, in <module>
    cost = backend.cost(compiled_circuit, n_shots=100, syntax_checker="H1-1SC")
  File "/home/alec/tmp/pytket-quantinuum/env/lib/python3.10/site-packages/pytket/extensions/quantinuum/backends/quantinuum.py", line 1045, in cost
    _ = backend.get_result(handle, use_websocket=use_websocket)
  File "/home/alec/tmp/pytket-quantinuum/env/lib/python3.10/site-packages/pytket/extensions/quantinuum/backends/quantinuum.py", line 970, in get_result
    backres = _convert_result(res, ppcirc, n_bits)
  File "/home/alec/tmp/pytket-quantinuum/env/lib/python3.10/site-packages/pytket/extensions/quantinuum/backends/quantinuum.py", line 1089, in _convert_result
    assert n_bits >= 0 and n_bits <= len(c_bits)
AssertionError
```

The same code run using the `develop` version produces a different, but clearly related, error:
```
  File "/home/alec/r/pytket-quantinuum/pytket/extensions/quantinuum/backends/quantinuum.py", line 1164, in cost
    _ = backend.get_result(handle, use_websocket=use_websocket)
  File "/home/alec/r/pytket-quantinuum/pytket/extensions/quantinuum/backends/quantinuum.py", line 1089, in get_result
    backres = _convert_result(res, ppcirc, n_bits, results_selection)
  File "/home/alec/r/pytket-quantinuum/pytket/extensions/quantinuum/backends/quantinuum.py", line 1228, in _convert_result
    stacked_array = [
  File "/home/alec/r/pytket-quantinuum/pytket/extensions/quantinuum/backends/quantinuum.py", line 1229, in <listcomp>
    [int(resultdict[name][i][-1 - ind]) for name, ind in results_selection]
  File "/home/alec/r/pytket-quantinuum/pytket/extensions/quantinuum/backends/quantinuum.py", line 1229, in <listcomp>
    [int(resultdict[name][i][-1 - ind]) for name, ind in results_selection]
KeyError: 'tk_SCRATCH_BIT'
```

### PR DESCRIPTION
**Pull request by cqc-alec**
Fixes #213 .

Integration tests running here: https://github.com/CQCL/pytket-quantinuum/actions/runs/6036474819 .

### CODE REVIEW

File: pytket/extensions/quantinuum/backends/quantinuum.py [hunk 1]
```diff
@@ -797,7 +795,9 @@ def process_circuits(
             else:
                 c0, ppcirc_rep = circ, None
             results_selection = []
-            for name, count in Counter(bit.reg_name for bit in c0.bits).items():
+            for name, count in Counter(
+                bit.reg_name for bit in c0.bits if not _is_scratch(bit)
```
Comment (yao-cqc) [inline pytket/extensions/quantinuum/backends/quantinuum.py:799]: Does this always ignore scratch bits, even if they are not erased?

Comment (cqc-alec) [reply pytket/extensions/quantinuum/backends/quantinuum.py:799]: It removes them all, which I assumed would be what we wanted. Can there be scratch bits that are actually used?

Comment (cqc-alec) [reply pytket/extensions/quantinuum/backends/quantinuum.py:799]: Yes I guess we cannot assume this...

Comment (cqc-alec) [reply pytket/extensions/quantinuum/backends/quantinuum.py]: Fixed I think. Integration tests running here: https://github.com/CQCL/pytket-quantinuum/actions/runs/6039214979 .

**Pull request merged**